File: src/semp_workflow/modules/client_profile.py
"""Client Profile module - client_profile.add, client_profile.delete."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient
from ..semp.helpers import check_name_length, clean_payload, coerce_bool, coerce_int, enc
from .base import BaseModule

logger = logging.getLogger(__name__)

BOOL_FIELDS = {
    "allowGuaranteedMsgSendEnabled",
    "allowGuaranteedMsgReceiveEnabled",
    "allowTransactedSessionsEnabled",
    "allowBridgeConnectionsEnabled",
}

INT_FIELDS = {
    "maxConnectionCountPerClientUsername",
    "maxEgressFlowCount",
    "maxIngressFlowCount",
    "maxSubscriptionCount",
}


def _build_profile_payload(args: dict) -> dict:
    payload = clean_payload(args)
    for key in BOOL_FIELDS:
        if key in payload:
            payload[key] = coerce_bool(payload[key])
    for key in INT_FIELDS:
        if key in payload:
            payload[key] = coerce_int(payload[key])
    return payload


class ClientProfileAdd(BaseModule):
    description = "Create a client profile on the message VPN. Skipped if the profile already exists."
    params = {
        "clientProfileName":                    {"type": "string",  "required": True,  "description": "Name of the client profile"},
        "allowGuaranteedMsgSendEnabled":         {"type": "boolean", "required": False, "description": "Allow clients to send guaranteed messages"},
        "allowGuaranteedMsgReceiveEnabled":      {"type": "boolean", "required": False, "description": "Allow clients to receive guaranteed messages"},
        "allowTransactedSessionsEnabled":        {"type": "boolean", "required": False, "description": "Allow clients to use transacted sessions"},
        "allowBridgeConnectionsEnabled":         {"type": "boolean", "required": False, "description": "Allow clients to use bridge connections"},
        "maxConnectionCountPerClientUsername":   {"type": "integer", "required": False, "description": "Maximum connections per client username (0 = unlimited)"},
        "maxEgressFlowCount":                    {"type": "integer", "required": False, "description": "Maximum number of egress flows per client"},
        "maxIngressFlowCount":                   {"type": "integer", "required": False, "description": "Maximum number of ingress flows per client"},
        "maxSubscriptionCount":                  {"type": "integer", "required": False, "description": "Maximum number of subscriptions per client"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        name = args.get("clientProfileName", "")
        if not name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: clientProfileName")

        if err := check_name_length("clientProfileName", name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"clientProfiles/{enc(name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking client profile: {e}")

        if exists:
            return ActionResult(ResultStatus.SKIPPED, f"Client profile '{name}' already exists")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would create client profile '{name}'")

        try:
            payload = _build_profile_payload(args)
            client.create("clientProfiles", payload)
            return ActionResult(
                ResultStatus.OK, f"Client profile '{name}' created"            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED, f"Failed to create client profile '{name}': {e}"
            )


class ClientProfileDelete(BaseModule):
    description = "Delete a client profile from the message VPN. Skipped if the profile does not exist."
    params = {
        "clientProfileName": {"type": "string", "required": True, "description": "Name of the client profile to delete"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        name = args.get("clientProfileName", "")
        if not name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: clientProfileName")

        if err := check_name_length("clientProfileName", name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"clientProfiles/{enc(name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking client profile: {e}")

        if not exists:
            return ActionResult(ResultStatus.SKIPPED, f"Client profile '{name}' does not exist")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would delete client profile '{name}'")

        try:
            client.delete(path)
            return ActionResult(
                ResultStatus.OK, f"Client profile '{name}' deleted"            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED, f"Failed to delete client profile '{name}': {e}"
            )


MODULES = {
    "client_profile.add": ClientProfileAdd,
    "client_profile.delete": ClientProfileDelete,
}

File: internal/modules/client_profile.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("client_profile.add", clientProfileAdd{})
	register("client_profile.delete", clientProfileDelete{})
	register("client_profile.update", clientProfileUpdate{})
}

var clientProfileBoolFields = []string{
	"allowGuaranteedMsgSendEnabled",
	"allowGuaranteedMsgReceiveEnabled",
	"allowTransactedSessionsEnabled",
	"allowBridgeConnectionsEnabled",
	"compressionEnabled",
}

var clientProfileIntFields = []string{
	"maxConnectionCountPerClientUsername",
	"maxEgressFlowCount",
	"maxIngressFlowCount",
	"maxSubscriptionCount",
}

func buildProfilePayload(args map[string]any) (map[string]any, error) {
	payload := semp.CleanPayload(args)
	coerceBoolFields(payload, clientProfileBoolFields...)
	if err := coerceIntFields(payload, clientProfileIntFields...); err != nil {
		return nil, err
	}
	return payload, nil
}

func clientProfileParams(nameDesc string) []ParamSpec {
	return []ParamSpec{
		{Name: "clientProfileName", Type: "string", Required: true, Description: nameDesc},
		{Name: "allowGuaranteedMsgSendEnabled", Type: "boolean", Description: "Allow clients to send guaranteed messages"},
		{Name: "allowGuaranteedMsgReceiveEnabled", Type: "boolean", Description: "Allow clients to receive guaranteed messages"},
		{Name: "allowTransactedSessionsEnabled", Type: "boolean", Description: "Allow clients to use transacted sessions"},
		{Name: "allowBridgeConnectionsEnabled", Type: "boolean", Description: "Allow clients to use bridge connections"},
		{Name: "compressionEnabled", Type: "boolean", Description: "Enable message compression for clients using this profile"},
		{Name: "maxConnectionCountPerClientUsername", Type: "integer", Description: "Maximum connections per client username (0 = unlimited)"},
		{Name: "maxEgressFlowCount", Type: "integer", Description: "Maximum number of egress flows per client"},
		{Name: "maxIngressFlowCount", Type: "integer", Description: "Maximum number of ingress flows per client"},
		{Name: "maxSubscriptionCount", Type: "integer", Description: "Maximum number of subscriptions per client"},
	}
}

type clientProfileAdd struct{}

func (clientProfileAdd) Description() string {
	return "Create a client profile on the message VPN. Skipped if the profile already exists."
}

func (clientProfileAdd) Params() []ParamSpec {
	return clientProfileParams("Name of the client profile")
}

func (clientProfileAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	name := argStr(args, "clientProfileName", "")
	if name == "" {
		return failed("Missing required arg: clientProfileName")
	}
	if err := semp.CheckNameLength("clientProfileName", name); err != "" {
		return failed(err)
	}

	path := "clientProfiles/" + semp.Enc(name)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking client profile: %s", err))
	}
	if exists {
		return skipped(fmt.Sprintf("Client profile '%s' already exists", name))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would create client profile '%s'", name))
	}

	payload, err := buildProfilePayload(args)
	if err != nil {
		return failed(err.Error())
	}
	if _, err := c.Create("clientProfiles", payload); err != nil {
		return failed(fmt.Sprintf("Failed to create client profile '%s': %s", name, err))
	}
	return ok(fmt.Sprintf("Client profile '%s' created", name))
}

type clientProfileDelete struct{}

func (clientProfileDelete) Description() string {
	return "Delete a client profile from the message VPN. Skipped if the profile does not exist."
}

func (clientProfileDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "clientProfileName", Type: "string", Required: true, Description: "Name of the client profile to delete"},
	}
}

func (clientProfileDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	name := argStr(args, "clientProfileName", "")
	if name == "" {
		return failed("Missing required arg: clientProfileName")
	}
	if err := semp.CheckNameLength("clientProfileName", name); err != "" {
		return failed(err)
	}

	path := "clientProfiles/" + semp.Enc(name)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking client profile: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("Client profile '%s' does not exist", name))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would delete client profile '%s'", name))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to delete client profile '%s': %s", name, err))
	}
	return ok(fmt.Sprintf("Client profile '%s' deleted", name))
}

type clientProfileUpdate struct{}

func (clientProfileUpdate) Description() string {
	return "Update attributes of an existing client profile. Fails if the profile does not exist."
}

func (clientProfileUpdate) Params() []ParamSpec {
	return clientProfileParams("Name of the client profile to update")
}

func (clientProfileUpdate) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	name := argStr(args, "clientProfileName", "")
	if name == "" {
		return failed("Missing required arg: clientProfileName")
	}
	if err := semp.CheckNameLength("clientProfileName", name); err != "" {
		return failed(err)
	}

	path := "clientProfiles/" + semp.Enc(name)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking client profile: %s", err))
	}
	if !exists {
		return failed(fmt.Sprintf("Client profile '%s' does not exist", name))
	}

	payload, err := buildProfilePayload(args)
	if err != nil {
		return failed(err.Error())
	}
	delete(payload, "clientProfileName")

	if len(payload) == 0 {
		return skipped("No fields to update")
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would update client profile '%s'", name))
	}

	if _, err := c.Update(path, payload); err != nil {
		return failed(fmt.Sprintf("Failed to update client profile '%s': %s", name, err))
	}
	return ok(fmt.Sprintf("Client profile '%s' updated", name))
}
